rustdoc-json: `#[target_feature]` incorrectly causes safe functions to be listed as `unsafe`

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
#[target_feature(enable = "sse2")]
pub fn compute(x: u32) -> u32 {
    x + 1
}
```

I expected to see this happen: rustdoc JSON should contain an entry for `compute` whose `header` field says:
```
"header": {
    "is_const": false,
    "is_unsafe": false,
    "is_async": false,
    "abi": "Rust"
},
```

Instead, this happened: the rustdoc JSON entry for `compute` incorrectly lists it as `unsafe`. Here's its entire item entry from a `format_version: 46` rustdoc JSON file:
```json5
"0": {
    "id": 0,
    "crate_id": 0,
    "name": "compute",
    "span": {
        "filename": "src/lib.rs",
        "begin": [
            2,
            1
        ],
        "end": [
            4,
            2
        ]
    },
    "visibility": "public",
    "docs": null,
    "links": {},
    "attrs": [
        "#[target_feature(enable = \"sse2\")]"
    ],
    "deprecation": null,
    "inner": {
        "function": {
            "sig": {
                "inputs": [
                    [
                        "x",
                        {
                            "primitive": "u32"
                        }
                    ]
                ],
                "output": {
                    "primitive": "u32"
                },
                "is_c_variadic": false
            },
            "generics": {
                "params": [],
                "where_predicates": []
            },
            "header": {
                "is_const": false,
                "is_unsafe": true,  # <--- BUG!
                "is_async": false,
                "abi": "Rust"
            },
            "has_body": true
        }
    }
},
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.89.0-nightly (45acf54ee 2025-06-16)
binary: rustc
commit-hash: 45acf54eea118ed27927282b5e0bfdcd80b7987c
commit-date: 2025-06-16
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.5
```

@rustbot label +A-rustdoc-json

cc @aDotInTheVoid 

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

They can be called safely in other `#[target_feature]` functions, unless they're marked as unsafe. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=1b0ae7d8b4e4f14574ebfb8b4ed7cf9c)

```rust
#[target_feature(enable = "sse2")]
pub fn safe_if_sse2(x: u32) -> u32 {
    x + 1
}

#[target_feature(enable = "sse2")]
pub unsafe fn unsafe_even_if_sse2(x: u32) -> u32 {
    x + 1
}

#[target_feature(enable = "sse2")]
pub fn entry() {
    safe_if_sse2(1); // Fine
    unsafe_even_if_sse2(1); // Error!
}
```

```
error[E0133]: call to unsafe function `unsafe_even_if_sse2` is unsafe and requires unsafe block
  --> src/lib.rs:14:5
   |
14 |     unsafe_even_if_sse2(1); // Error!
   |     ^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
```

--- a/src/librustdoc/json/conversions.rs
+++ b/src/librustdoc/json/conversions.rs
@@ -9,6 +9,7 @@ use rustc_ast::ast;
 use rustc_attr_data_structures::{self as attrs, DeprecatedSince};
 use rustc_hir::def::CtorKind;
 use rustc_hir::def_id::DefId;
+use rustc_hir::{HeaderSafety, Safety};
 use rustc_metadata::rendered_const;
 use rustc_middle::{bug, ty};
 use rustc_span::{Pos, kw, sym};
@@ -383,10 +384,22 @@ impl FromClean<clean::Union> for Union {
 
 impl FromClean<rustc_hir::FnHeader> for FunctionHeader {
     fn from_clean(header: &rustc_hir::FnHeader, renderer: &JsonRenderer<'_>) -> Self {
+        let is_unsafe = match header.safety {
+            HeaderSafety::SafeTargetFeatures => {
+                // The type system's internal implementation details consider
+                // safe functions with the `#[target_feature]` attribute to be analogous
+                // to unsafe functions: `header.is_unsafe()` returns `true` for them.
+                // For rustdoc, this isn't the right decision, so we explicitly return `false`.
+                // Context: https://github.com/rust-lang/rust/issues/142655
+                false
+            }
+            HeaderSafety::Normal(Safety::Safe) => false,
+            HeaderSafety::Normal(Safety::Unsafe) => true,
+        };
         FunctionHeader {
             is_async: header.is_async(),
             is_const: header.is_const(),
-            is_unsafe: header.is_unsafe(),
+            is_unsafe,
             abi: header.abi.into_json(renderer),
         }
     }

--- a/tests/rustdoc-json/attrs/target_feature.rs
+++ b/tests/rustdoc-json/attrs/target_feature.rs
@@ -1,17 +1,40 @@
 //@ only-x86_64
 
 //@ is "$.index[?(@.name=='test1')].attrs" '["#[target_feature(enable=\"avx\")]"]'
+//@ is "$.index[?(@.name=='test1')].inner.function.header.is_unsafe" false
 #[target_feature(enable = "avx")]
 pub fn test1() {}
 
 //@ is "$.index[?(@.name=='test2')].attrs" '["#[target_feature(enable=\"avx\", enable=\"avx2\")]"]'
+//@ is "$.index[?(@.name=='test1')].inner.function.header.is_unsafe" false
 #[target_feature(enable = "avx,avx2")]
 pub fn test2() {}
 
 //@ is "$.index[?(@.name=='test3')].attrs" '["#[target_feature(enable=\"avx\", enable=\"avx2\")]"]'
+//@ is "$.index[?(@.name=='test1')].inner.function.header.is_unsafe" false
 #[target_feature(enable = "avx", enable = "avx2")]
 pub fn test3() {}
 
 //@ is "$.index[?(@.name=='test4')].attrs" '["#[target_feature(enable=\"avx\", enable=\"avx2\", enable=\"avx512f\")]"]'
+//@ is "$.index[?(@.name=='test1')].inner.function.header.is_unsafe" false
 #[target_feature(enable = "avx", enable = "avx2,avx512f")]
 pub fn test4() {}
+
+//@ is "$.index[?(@.name=='test_unsafe_fn')].attrs" '["#[target_feature(enable=\"avx\")]"]'
+//@ is "$.index[?(@.name=='test_unsafe_fn')].inner.function.header.is_unsafe" true
+#[target_feature(enable = "avx")]
+pub unsafe fn test_unsafe_fn() {}
+
+pub struct Example;
+
+impl Example {
+    //@ is "$.index[?(@.name=='safe_assoc_fn')].attrs" '["#[target_feature(enable=\"avx\")]"]'
+    //@ is "$.index[?(@.name=='safe_assoc_fn')].inner.function.header.is_unsafe" false
+    #[target_feature(enable = "avx")]
+    pub fn safe_assoc_fn() {}
+
+    //@ is "$.index[?(@.name=='unsafe_assoc_fn')].attrs" '["#[target_feature(enable=\"avx\")]"]'
+    //@ is "$.index[?(@.name=='unsafe_assoc_fn')].inner.function.header.is_unsafe" true
+    #[target_feature(enable = "avx")]
+    pub unsafe fn unsafe_assoc_fn() {}
+}

